<a href="https://colab.research.google.com/github/10zhu/GoldMiner/blob/main/8_EDA_Tutorial_ipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exploratory Data Analysis (EDA) Tutorial

## Introduction

### Meaning of Exploratory Data Analysis
In this tutorial, you'll learn how to use Pandas to perform the basic exploratory data analysis (EDA). EDA refers to any analysis done to understand the basic structure, characteristics, and quality of a dataset. This includes efforts to measure spread and center of numerical and categorical data, and efforts to identify outliers, misentered values, and missing values, as well as any basic investigation done to test basic assumptions about the data. Typically, EDA incorporates some amount of data visualization. 

### Purpose of EDA
Whether you’re compiling a database for a web application or training a deep learning model, EDA should be the first thing you do after data collection. Real datasets are almost never perfect. You’ll often find typos, misentered values, and chunks of missing data. And in most projects, some of your assumptions about how to interpret the values are probably incorrect. EDA allows you to identify these problems and incorrect assumptions early before you bake them into your model, application, or processing pipeline.

### EDA for the project
In this project, performing EDA will give you a sense of how much cleaning you’ll have to do to prepare the data for ingestion. It will also help you gauge how long basic operations on your data will take to execute. And it will reveal whether the contents of the dataset match your assumptions about how the data was collected and the fields were defined. 

All of this information will help you estimate how useful a dataset will be to your application and how difficult it will be to work with. Ultimately, this will inform whether you should commit to using a particular dataset or search for an alternative instead. Therefore, it’s essential that you perform some EDA before submitting your project proposal, so you’re not stuck with a bad dataset when it’s more difficult to change course later. 

### Tutorial Overview
This tutorial covers the EDA you'll need to perform when choosing datasets:

*  Data importation
*  Initial exploration
*  Missing value detection
*  Misentered value detection
*  Distribution examination

To edit and run the code throughout the tutorial, open the notebook in "playground mode" using the button in the upper right corner.

After going through the tutorial, try the accompanying exercises to practice what you learned. 

In [ ]:
import pandas as pd
import numpy as np


## Data importation


For this example, we'll work with a [subset of U.S. census data](https://archive.ics.uci.edu/ml/datasets/Adult) maintained by the UCI Machine Learning Repository. [The Repository](https://archive.ics.uci.edu/ml/index.php) contains many other free, easily downloadable datasets that might be useful for your project , so be sure to check it out when you're searching for datasets later on!

To import the data, first copy the dataset into your Drive by following [this link](https://drive.google.com/open?id=1WB60Q6VJYyjbE8h8WED2mfi8Ae8LzoRU) and clicking on the "Add shortcut to Drive" link in the upper-right corner. Remember where your store it in your drive. 


Next, mount your Google Drive to the notebook by running the code cell below. This will allow you to import any files in your drive to the Python Runtime Environment that the notebook is using. When you run the code cell, Google Drive may ask you to re-login to your account and paste a verification code into a popup below the cell.


In [ ]:
prefix = '/content/drive'
from google.colab import drive
drive.mount(prefix, force_remount=True)

Mounted at /content/drive


Now, you should see a directory named "drive" when you click the Files icon on the left. Expand this directory and navigate to the location you saved the census data. Control-click (right-click for PC) on the file and select "copy path". Then, paste the path into `file_path` variable in the cell below and run the cell. 

In [ ]:
file_path = '/content/drive/MyDrive/CIS550/anime.csv' # YOUR PATH TO THE CENSUS SUBSET HERE e.g. '/content/drive/My Drive/CIS550/adult.data'

The dataset is stored as a CSV, so we'll load it into a `DataFrame` using `pandas.read_csv` in the code cell below. `Pandas` also defines a number of other [IO functions](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html) that load the contents of common file types directly into `DataFrames`.

In [ ]:
census = pd.read_csv(file_path)
census.head()

anime_id                               title   type  score  scored_by  \
0      5114    Fullmetal Alchemist: Brotherhood     tv   9.13    1871705   
1     11061              Hunter x Hunter (2011)     tv   9.04    1509622   
2     38524  Shingeki no Kyojin Season 3 Part 2     tv   9.07    1329500   
3      9253                         Steins;Gate     tv   9.08    1252286   
4     28851                      Koe no Katachi  movie   8.95    1398608   

            status  episodes  start_date    end_date        source  ...  \
0  finished_airing      64.0  2009-04-05  2010-07-04         manga  ...   
1  finished_airing     148.0  2011-10-02  2014-09-24         manga  ...   
2  finished_airing      10.0  2019-04-29  2019-07-01         manga  ...   
3  finished_airing      24.0  2011-04-06  2011-09-14  visual_novel  ...   
4  finished_airing       1.0  2016-09-17  2016-09-17         manga  ...   

                                           producers  \
0  ['Aniplex', 'Square Enix', 'Mainichi Broadcast...   
1   ['VAP', 'Nippon Television Network', 'Shueisha']   
2  ['Production I.G', 'Dentsu', 'Mainichi Broadca...   
3  ['Frontier Works', 'Media Factory', 'Movic', '...   
4  ['Shochiku', 'Pony Canyon', 'Kodansha', 'ABC A...   

                              licensors  \
0  ['Funimation', 'Aniplex of America']   
1                         ['VIZ Media']   
2                        ['Funimation']   
3                        ['Funimation']   
4          ['Eleven Arts', 'NYAV Post']   

                                            synopsis  \
0  After a horrific alchemy experiment goes wrong...   
1  Hunters devote themselves to accomplishing haz...   
2  Seeking to restore humanity's diminishing hope...   
3  Eccentric scientist Rintarou Okabe has a never...   
4  As a wild youth, elementary school student Sho...   

                                          background  \
0                                                NaN   
1                                                NaN   
2  Shingeki no Kyojin adapts content from volumes...   
3  Steins;Gate is based on 5pb. and Nitroplus' re...   
4  Winner of the Excellence Award on the 20th Jap...   

                                        main_picture  \
0  https://cdn.myanimelist.net/images/anime/1223/...   
1  https://cdn.myanimelist.net/images/anime/1337/...   
2  https://cdn.myanimelist.net/images/anime/1517/...   
3  https://cdn.myanimelist.net/images/anime/5/731...   
4  https://cdn.myanimelist.net/images/anime/1122/...   

                                                 url  \
0  https://myanimelist.net/anime/5114/Fullmetal_A...   
1  https://myanimelist.net/anime/11061/Hunter_x_H...   
2  https://myanimelist.net/anime/38524/Shingeki_n...   
3     https://myanimelist.net/anime/9253/Steins_Gate   
4  https://myanimelist.net/anime/28851/Koe_no_Kat...   

                                   trailer_url  \
0  https://www.youtube.com/watch?v=--IcmZkvL0Q   
1  https://www.youtube.com/watch?v=D9iTQRB4XRk   
2  https://www.youtube.com/watch?v=hKHepjfj5Tw   
3  https://www.youtube.com/watch?v=27OZc-ku6is   
4  https://www.youtube.com/watch?v=XBNWo25izJ8   

                      title_english              title_japanese  \
0  Fullmetal Alchemist: Brotherhood  鋼の錬金術師 FULLMETAL ALCHEMIST   
1                   Hunter x Hunter    HUNTER×HUNTER（ハンター×ハンター）   
2   Attack on Titan Season 3 Part 2        進撃の巨人 Season3 Part.2   
3                       Steins;Gate                 STEINS;GATE   
4                    A Silent Voice                         聲の形   

                                      title_synonyms  
0  ['Hagane no Renkinjutsushi: Fullmetal Alchemis...  
1                                     ['HxH (2011)']  
2                                                 []  
3                                                 []  
4                             ['The Shape of Voice']  

[5 rows x 39 columns]

## Initial Exploration
Before thinking about cleaning, outlier detection, or anything else, we need to have a basic sense of what the dataset contains. The dataframe object defines a number of functions and properties that serve this purpose:


*   **DataFrame.shape**: Property that gives the gives the number of rows and columns in the DataFrame
*   **DataFrame.size**: Property that gives the total number of elements in the DataFrame
*   **DataFrame.columns**: Property that gives names of the DataFrame's columns
*   **DataFrame.dtypes**: Property that gives the data type of each column
*   **DataFrame.head()/.tail()**: Function that returns the first/last few rows of the DataFrame
*   **DataFrame.memory_usage()**: Function that returns the memory usage for each column of the DataFrame in bytes

Let's take a closer look at our DataFrame using `DataFrame.head()` first. 

In [ ]:
census

anime_id                                           title   type  score  \
0          5114                Fullmetal Alchemist: Brotherhood     tv   9.13   
1         11061                          Hunter x Hunter (2011)     tv   9.04   
2         38524              Shingeki no Kyojin Season 3 Part 2     tv   9.07   
3          9253                                     Steins;Gate     tv   9.08   
4         28851                                  Koe no Katachi  movie   8.95   
...         ...                                             ...    ...    ...   
24160     52470                                      STARTLINER  music    NaN   
24161     52474  Blue Archive: 1.5 Shuunenkinen Short Animation    ona    NaN   
24162     52475                                  Ryuujin Kappei    ova    NaN   
24163     52476                              Yeongnihan neoguri    NaN    NaN   
24164     52477                     Dab-eul chaj-eun du sonyeon    NaN    NaN   

       scored_by           status  episodes  start_date    end_date  \
0        1871705  finished_airing      64.0  2009-04-05  2010-07-04   
1        1509622  finished_airing     148.0  2011-10-02  2014-09-24   
2        1329500  finished_airing      10.0  2019-04-29  2019-07-01   
3        1252286  finished_airing      24.0  2011-04-06  2011-09-14   
4        1398608  finished_airing       1.0  2016-09-17  2016-09-17   
...          ...              ...       ...         ...         ...   
24160          0  finished_airing       1.0  2018-07-25  2018-07-25   
24161          0  finished_airing       1.0  2022-07-16         NaN   
24162          0  finished_airing       1.0  1991-04-21  1991-04-21   
24163          0  finished_airing      57.0  1988-01-01  2007-01-01   
24164          0  finished_airing       1.0  2015-01-01  2015-01-01   

             source  ...                                          producers  \
0             manga  ...  ['Aniplex', 'Square Enix', 'Mainichi Broadcast...   
1             manga  ...   ['VAP', 'Nippon Television Network', 'Shueisha']   
2             manga  ...  ['Production I.G', 'Dentsu', 'Mainichi Broadca...   
3      visual_novel  ...  ['Frontier Works', 'Media Factory', 'Movic', '...   
4             manga  ...  ['Shochiku', 'Pony Canyon', 'Kodansha', 'ABC A...   
...             ...  ...                                                ...   
24160          game  ...                                                 []   
24161           NaN  ...                                                 []   
24162      original  ...                                                 []   
24163      original  ...                                                 []   
24164           NaN  ...                                                 []   

                                  licensors  \
0      ['Funimation', 'Aniplex of America']   
1                             ['VIZ Media']   
2                            ['Funimation']   
3                            ['Funimation']   
4              ['Eleven Arts', 'NYAV Post']   
...                                     ...   
24160                                    []   
24161                                    []   
24162                                    []   
24163                                    []   
24164                                    []   

                                                synopsis  \
0      After a horrific alchemy experiment goes wrong...   
1      Hunters devote themselves to accomplishing haz...   
2      Seeking to restore humanity's diminishing hope...   
3      Eccentric scientist Rintarou Okabe has a never...   
4      As a wild youth, elementary school student Sho...   
...                                                  ...   
24160  Music video for the main theme of ONGEKI, "STA...   
24161                                                NaN   
24162  Kappei got drown into a strange picture book. ...   
24163  Clever Raccoon Dog is a North Korean animated ...   
24164  

In [ ]:
census['score'].isna().sum()

9893

### Fixing Column Names
The index looks fine, but the column names don't look correct. The names appear to match the contents of columns in many cases. As a result, we can conclude that `pandas.read_csv` expected the first row of the `.data` file to be a list of labels, but it turned out to be the first data point. 

[The documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) for `pandas.read_csv` indicates how we can override this default behavior. According to the documentation, we can set `header` to `None` and pass a list of column names to the `names` argument to treat the first row as a row of data, rather than column names. Let's do that in the chunk below, then re-run `DataFrame.head` to make sure it worked as we expect. 

In [ ]:
col_names = ['source', 'title','episodes',  'score', 'total_duration',
             'favorites', 'genres', 'start_date']  # names taken from the UCI ML Repo homepage for the data
df = census[col_names]
df.head()

source                               title  episodes  score  \
0         manga    Fullmetal Alchemist: Brotherhood      64.0   9.13   
1         manga              Hunter x Hunter (2011)     148.0   9.04   
2         manga  Shingeki no Kyojin Season 3 Part 2      10.0   9.07   
3  visual_novel                         Steins;Gate      24.0   9.08   
4         manga                      Koe no Katachi       1.0   8.95   

    total_duration  favorites                                       genres  \
0  1 days 01:57:20     204645  ['Action', 'Adventure', 'Drama', 'Fantasy']   
1  2 days 10:15:16     185178           ['Action', 'Adventure', 'Fantasy']   
2  0 days 03:59:10      51931                          ['Action', 'Drama']   
3  0 days 09:44:00     173088              ['Drama', 'Sci-Fi', 'Suspense']   
4  0 days 02:10:03      77431                                    ['Drama']   

   start_date  
0  2009-04-05  
1  2011-10-02  
2  2019-04-29  
3  2011-04-06  
4  2016-09-17

In [ ]:
df[['total_duration_date','data', 'time']] = df.total_duration.str.split(" ",expand=True)
df = df.drop(['data'], axis=1)
df.head()

<ipython-input-8-cb2cf89bc3a2>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[['total_duration_date','data', 'time']] = df.total_duration.str.split(" ",expand=True)
<ipython-input-8-cb2cf89bc3a2>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[['total_duration_date','data', 'time']] = df.total_duration.str.split(" ",expand=True)
<ipython-input-8-cb2cf89bc3a2>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = valu

source                               title  episodes  score  \
0         manga    Fullmetal Alchemist: Brotherhood      64.0   9.13   
1         manga              Hunter x Hunter (2011)     148.0   9.04   
2         manga  Shingeki no Kyojin Season 3 Part 2      10.0   9.07   
3  visual_novel                         Steins;Gate      24.0   9.08   
4         manga                      Koe no Katachi       1.0   8.95   

    total_duration  favorites                                       genres  \
0  1 days 01:57:20     204645  ['Action', 'Adventure', 'Drama', 'Fantasy']   
1  2 days 10:15:16     185178           ['Action', 'Adventure', 'Fantasy']   
2  0 days 03:59:10      51931                          ['Action', 'Drama']   
3  0 days 09:44:00     173088              ['Drama', 'Sci-Fi', 'Suspense']   
4  0 days 02:10:03      77431                                    ['Drama']   

   start_date total_duration_date      time  
0  2009-04-05                   1  01:57:20  
1  2011-10-02                   2  10:15:16  
2  2019-04-29                   0  03:59:10  
3  2011-04-06                   0  09:44:00  
4  2016-09-17                   0  02:10:03

In [ ]:
# df['total_duration_date'] = df['total_duration_date'].fillna(0)

In [ ]:
# df.dropna()
df = df.dropna(subset=['total_duration_date'])

In [ ]:
df

source                                           title  episodes  \
0             manga                Fullmetal Alchemist: Brotherhood      64.0   
1             manga                          Hunter x Hunter (2011)     148.0   
2             manga              Shingeki no Kyojin Season 3 Part 2      10.0   
3      visual_novel                                     Steins;Gate      24.0   
4             manga                                  Koe no Katachi       1.0   
...             ...                                             ...       ...   
24160          game                                      STARTLINER       1.0   
24161           NaN  Blue Archive: 1.5 Shuunenkinen Short Animation       1.0   
24162      original                                  Ryuujin Kappei       1.0   
24163      original                              Yeongnihan neoguri      57.0   
24164           NaN                     Dab-eul chaj-eun du sonyeon       1.0   

       score   total_duration  favorites  \
0       9.13  1 days 01:57:20     204645   
1       9.04  2 days 10:15:16     185178   
2       9.07  0 days 03:59:10      51931   
3       9.08  0 days 09:44:00     173088   
4       8.95  0 days 02:10:03      77431   
...      ...              ...        ...   
24160    NaN  0 days 00:03:31          0   
24161    NaN  0 days 00:09:13          0   
24162    NaN  0 days 00:15:00          0   
24163    NaN  0 days 12:21:00          0   
24164    NaN  0 days 00:17:41          0   

                                            genres  start_date  \
0      ['Action', 'Adventure', 'Drama', 'Fantasy']  2009-04-05   
1               ['Action', 'Adventure', 'Fantasy']  2011-10-02   
2                              ['Action', 'Drama']  2019-04-29   
3                  ['Drama', 'Sci-Fi', 'Suspense']  2011-04-06   
4                                        ['Drama']  2016-09-17   
...                                            ...         ...   
24160                                           []  2018-07-25   
24161                        ['Action', 'Fantasy']  2022-07-16   
24162           ['Action', 'Adventure', 'Fantasy']  1991-04-21   
24163                                           []  1988-01-01   
24164                                           []  2015-01-01   

      total_duration_date      time  
0                       1  01:57:20  
1                       2  10:15:16  
2                       0  03:59:10  
3                       0  09:44:00  
4                       0  02:10:03  
...                   ...       ...  
24160                   0  00:03:31  
24161                   0  00:09:13  
24162                   0  00:15:00  
24163                   0  12:21:00  
24164                   0  00:17:41  

[23002 rows x 10 columns]

In [ ]:
df.total_duration_date = df['total_duration_date'].astype('int')
print(df.dtypes)

source                  object
title                   object
episodes               float64
score                  float64
total_duration          object
favorites                int64
genres                  object
start_date              object
total_duration_date      int64
time                    object
dtype: object


In [ ]:
df[['hours','minutes', 'seconds']] = df.time.str.split(":",expand=True)
df = df.drop(['seconds'], axis=1)
df.hours = df['hours'].astype('int')
df.minutes = df['minutes'].astype('int')
sum_col = df['total_duration_date']*24*60 + df['hours']*60 + df['minutes']
df['dur'] = sum_col
df.head()

source                               title  episodes  score  \
0         manga    Fullmetal Alchemist: Brotherhood      64.0   9.13   
1         manga              Hunter x Hunter (2011)     148.0   9.04   
2         manga  Shingeki no Kyojin Season 3 Part 2      10.0   9.07   
3  visual_novel                         Steins;Gate      24.0   9.08   
4         manga                      Koe no Katachi       1.0   8.95   

    total_duration  favorites                                       genres  \
0  1 days 01:57:20     204645  ['Action', 'Adventure', 'Drama', 'Fantasy']   
1  2 days 10:15:16     185178           ['Action', 'Adventure', 'Fantasy']   
2  0 days 03:59:10      51931                          ['Action', 'Drama']   
3  0 days 09:44:00     173088              ['Drama', 'Sci-Fi', 'Suspense']   
4  0 days 02:10:03      77431                                    ['Drama']   

   start_date  total_duration_date      time  hours  minutes   dur  
0  2009-04-05                    1  01:57:20      1       57  1557  
1  2011-10-02                    2  10:15:16     10       15  3495  
2  2019-04-29                    0  03:59:10      3       59   239  
3  2011-04-06                    0  09:44:00      9       44   584  
4  2016-09-17                    0  02:10:03      2       10   130

In [ ]:
df = df.drop(['total_duration_date','total_duration','hours','time','minutes'], axis=1)

In [ ]:
df.head()

source                               title  episodes  score  \
0         manga    Fullmetal Alchemist: Brotherhood      64.0   9.13   
1         manga              Hunter x Hunter (2011)     148.0   9.04   
2         manga  Shingeki no Kyojin Season 3 Part 2      10.0   9.07   
3  visual_novel                         Steins;Gate      24.0   9.08   
4         manga                      Koe no Katachi       1.0   8.95   

   favorites                                       genres  start_date   dur  
0     204645  ['Action', 'Adventure', 'Drama', 'Fantasy']  2009-04-05  1557  
1     185178           ['Action', 'Adventure', 'Fantasy']  2011-10-02  3495  
2      51931                          ['Action', 'Drama']  2019-04-29   239  
3     173088              ['Drama', 'Sci-Fi', 'Suspense']  2011-04-06   584  
4      77431                                    ['Drama']  2016-09-17   130

In [ ]:
df.rename(columns = {'dur': 'total_duration'},inplace = True)

Now, we have meaningful column names. And as a sanity check, you can ensure that the first row of the data in our new dataframe matches the column names in the old dataframe, so you know we didn't accidently throw out the first row.

In [ ]:
df.head()

source                               title  episodes  score  \
0         manga    Fullmetal Alchemist: Brotherhood      64.0   9.13   
1         manga              Hunter x Hunter (2011)     148.0   9.04   
2         manga  Shingeki no Kyojin Season 3 Part 2      10.0   9.07   
3  visual_novel                         Steins;Gate      24.0   9.08   
4         manga                      Koe no Katachi       1.0   8.95   

   favorites                                       genres  start_date  \
0     204645  ['Action', 'Adventure', 'Drama', 'Fantasy']  2009-04-05   
1     185178           ['Action', 'Adventure', 'Fantasy']  2011-10-02   
2      51931                          ['Action', 'Drama']  2019-04-29   
3     173088              ['Drama', 'Sci-Fi', 'Suspense']  2011-04-06   
4      77431                                    ['Drama']  2016-09-17   

   total_duration  
0            1557  
1            3495  
2             239  
3             584  
4             130

In [ ]:
avg = df["score"].mean()
print(avg)

6.458897832380145


In [ ]:
df['score'] = df['score'].fillna(avg)

In [ ]:
df

source                                           title  episodes  \
0             manga                Fullmetal Alchemist: Brotherhood      64.0   
1             manga                          Hunter x Hunter (2011)     148.0   
2             manga              Shingeki no Kyojin Season 3 Part 2      10.0   
3      visual_novel                                     Steins;Gate      24.0   
4             manga                                  Koe no Katachi       1.0   
...             ...                                             ...       ...   
24160          game                                      STARTLINER       1.0   
24161           NaN  Blue Archive: 1.5 Shuunenkinen Short Animation       1.0   
24162      original                                  Ryuujin Kappei       1.0   
24163      original                              Yeongnihan neoguri      57.0   
24164           NaN                     Dab-eul chaj-eun du sonyeon       1.0   

          score  favorites                                       genres  \
0      9.130000     204645  ['Action', 'Adventure', 'Drama', 'Fantasy']   
1      9.040000     185178           ['Action', 'Adventure', 'Fantasy']   
2      9.070000      51931                          ['Action', 'Drama']   
3      9.080000     173088              ['Drama', 'Sci-Fi', 'Suspense']   
4      8.950000      77431                                    ['Drama']   
...         ...        ...                                          ...   
24160  6.458898          0                                           []   
24161  6.458898          0                        ['Action', 'Fantasy']   
24162  6.458898          0           ['Action', 'Adventure', 'Fantasy']   
24163  6.458898          0                                           []   
24164  6.458898          0                                           []   

       start_date  total_duration  
0      2009-04-05            1557  
1      2011-10-02            3495  
2      2019-04-29             239  
3      2011-04-06             584  
4      2016-09-17             130  
...           ...             ...  
24160  2018-07-25               3  
24161  2022-07-16               9  
24162  1991-04-21              15  
24163  1988-01-01             741  
24164  2015-01-01              17  

[23002 rows x 8 columns]

In [ ]:
df.to_csv('/content/drive/MyDrive/anime_out.csv')  


### Early Data Interpretation
Let's use the output of `census.head()` to draw conclusions and characterize our uncertainties about how to interpret each column. 


The intended meanings of many of the columns are easy to infer: `age`, `education`, `maritial-status`, `relationship`, `race`, `sex`, `hours-per-week`, and `native-country`. But the meanings of a few columns (like `fnlwgt`, `education-num`, `capital-gain`, and `capital-loss`) aren't immediately obvious.

And at this stage, we should have many unanswered questions related to interpretation. For example:  
*   How many possible values can each categorical variable take on? How should we interpret each of these?
*   Are `education` and `education-num` redundant? 
*   Can a person have non-zero `capital-gain` and non-zero `capital-loss` simultaneously? What would this mean?

### Dataset Size Measurement

Let's use a couple of the other functions we discussed above to get a better sense of how large the dataset and what it contains. 

In [ ]:
census.dtypes

anime_id              int64
title                object
type                 object
score               float64
scored_by             int64
status               object
episodes            float64
start_date           object
end_date             object
source               object
members               int64
favorites             int64
episode_duration     object
total_duration       object
rating               object
sfw                    bool
approved               bool
created_at           object
updated_at           object
start_year          float64
start_season         object
real_start_date      object
real_end_date        object
broadcast_day        object
broadcast_time       object
genres               object
themes               object
demographics         object
studios              object
producers            object
licensors            object
synopsis             object
background           object
main_picture         object
url                  object
trailer_url         

In [ ]:
census.shape  # (first number counts rows)

(24165, 39)

In [ ]:
usage_by_col = census.memory_usage(deep=True) # returns memory usage for each column
total_usage = usage_by_col.sum()
mbs = total_usage / 1e6 # convert to megabytes 
mbs

61.812172

The outputs of the above commands tell us that the dataframe consists of about 33 thousand rows, 9 string columns, and 6 integer columns. 

We also now know that the `DataFrame` occupies about 21 megabtyes of RAM. To put this number in context, have a look at the RAM usage in the upper right corner. We have a lot of unused RAM capacity (about 12 GBs), so we know we can continue to work with this dataset without risk of crashing our environment.  

## Missing Value Detection
Now, let's find out which (if any) rows and columns contain missing values. 

Typically in a `DataFrame`, `np.NaN` represents a missing value. We will use the following two functions in combination to find out which columns have missing values: 
*   **`DataFrame.isna()`**: Returns a `DataFrame` where every value is a boolean that represents whether the value was `np.NaN` in the input `DataFrame`
*   **`DataFrame.any()`**: Returns a `Series` that indicates whether each column contained at least one `True` boolean. 

In [ ]:
census.isnull().any()

anime_id            False
title               False
type                 True
score                True
scored_by           False
status              False
episodes             True
start_date           True
end_date             True
source               True
members             False
favorites           False
episode_duration     True
total_duration       True
rating               True
sfw                 False
approved            False
created_at          False
updated_at          False
start_year           True
start_season         True
real_start_date      True
real_end_date        True
broadcast_day        True
broadcast_time       True
genres              False
themes              False
demographics        False
studios             False
producers           False
licensors           False
synopsis             True
background           True
main_picture         True
url                 False
trailer_url          True
title_english        True
title_japanese       True
title_synony

This suggests there are no missing values. But we assumed that the dataset doesn't use a placeholder value for missing values in the text column. Let's check this assumption by printing the unique values for each text column:

In [ ]:
# for col in census.columns:  # iterate over all columns
#   if census[col].dtype not in [np.int64, np.float64]:  # ignore integer and float columns
#     # print a header, all unique values, and a separator
#     print('Unique Values for {}'.format(col))  
#     print(census[col].unique())   
#     print('')

An inspection of the unique values for each column indicates `workclass`, `occupation`, and `native-country` all contain at least one " ?". Since there's no question mark country, this is probably the missing value placeholder we were looking for. 

Let's find out what portion of values are missing from each of those affected columns.  

In [ ]:
# missing = census[['native-country', 'workclass', 'occupation']].copy()  # subset columns
#  Redefine each column as a boolean indicating whether the original value was `?`
# for col in missing.columns:
#   missing[col] = missing[col] == " ?"
# missing.sum()  # counts the number of 'True' values in each column (Treats True as 1, False as 0)

In the worst case, this means 4262 of 32k rows have at least one missing value. Let's find out exactly how many rows are affected.

By passing `axis='columns'` to `DataFrame.any()`, we check whether each row contains at least 1 `True` value, then we can count the rows with true values using `Series.sum`:

In [ ]:
missing.any(axis='columns').sum()

NameError: ignored

As a fraction of all rows:

In [ ]:
(missing.any(axis='columns').sum()) / missing.shape[0]

If you were considering using this dataset for your project, you would want to think about whether and how your application could tolerate this amount of missing data. 

Depending on your application, you may want throw out the rows with missing values, use some placeholder value, or abandon the dataset altogether. 

## Misentered Value Detection

Now, let's see if the dataset contains any obviously misentered values. 

The unique values of the text data we printed above suggest there aren't any misentered text values. But we still need to investigate the numeric columns. Lets use the `DataFrame.max()` and `DataFrame.min()` functions to check for any obviously implausible values. 

In [ ]:
census.max()

In [ ]:
census.min()

Nothing in the min and max outputs above indicates there are obviously misentered values. For example, we would have had a problem if the minimum age were less than 0 or the max were greater than 130. 

Even though we didn't find any misentered values, the min and max data still allow us to draw some new conclusions about some of the columns. For example: 
*   `education-num` is defined on range 1 to 16. This suggests it may be the number of years of school completed through undergrad. 
*   `capital-gain` and `capital-loss` are nonnegative, but we still don't know if they can be positive simulataneously. 


## Distribution Examination

Finally, let's take a closer look at the distributions of a couple features. 

##### Numerical Distributions
Pandas allows us to easily obtain summary statistics or generate plots to learn more about numerical distributions.

`DataFrame.describe()` and `Series.describe()` give a few useful statistics that summarize center and spread. For example:

In [ ]:
census['age'].describe()

We can also generate a histogram to visualize the shape of the distribution:

In [ ]:
census['age'].plot(kind='hist')

Or a box plot to visualize spread and identify outliers:

In [ ]:
census['age'].plot(kind='box')

##### Categorical Distributions
We can learn more about the distribution of a categorical variable by generating a bar plot of category frequencies. For example: 

In [ ]:
census['maritial-status'].value_counts().plot(kind='bar')

Or a pie plot:

In [ ]:
census['maritial-status'].value_counts().plot(kind='pie')

## Exercises
Check out [these exercises](https://drive.google.com/open?id=11K6Ivr4viUKCKblF64qG2kU1KNodnIxY) to practice the EDA techniques you learned above!